File: java/app/src/main/java/com/pgssoft/mvvm/views/activities/MainActivity.java
package com.pgssoft.mvvm.views.activities;

import android.databinding.DataBindingUtil;
import android.databinding.ObservableField;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;

import com.pgssoft.mvvm.MVVMApplication;
import com.pgssoft.mvvm.R;
import com.pgssoft.mvvm.databinding.ActivityMainBinding;
import com.pgssoft.mvvm.model.database.Rate;
import com.pgssoft.mvvm.services.ServiceProvider;
import com.pgssoft.mvvm.viewmodels.MainActivityViewModel;
import com.pgssoft.mvvm.viewmodels.interfaces.IMainActivityAccess;
import com.pgssoft.mvvm.views.adapters.RatesAdapter;
import com.pgssoft.mvvm.views.adapters.interfaces.handlers.RateItemHandler;
import com.pgssoft.mvvm.views.adapters.interfaces.providers.BaseItemProvider;
import com.pgssoft.mvvm.views.adapters.interfaces.providers.RatesAdapterProvider;

public class MainActivity extends BaseViewModelActivity<MainActivityViewModel>
        implements IMainActivityAccess, RatesAdapterProvider {
    private ActivityMainBinding binding;
    private ObservableField<RatesAdapter> ratesAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        ratesAdapter = new ObservableField<>();
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));

        ServiceProvider serviceProvider = MVVMApplication.getServiceProvider();
        viewModel = new MainActivityViewModel(this, serviceProvider.getApiService(),
                serviceProvider.getRepository());

        binding.setViewModel(viewModel);
        binding.setAdapterProvider(this);

        viewModel.onInfrastructureReady();
    }

    @Override
    public void displayRates(BaseItemProvider<Rate> ratesProvider, RateItemHandler rateItemHandler) {
        ratesAdapter.set(new RatesAdapter(this, ratesProvider, rateItemHandler));
    }

    @Override
    public void openRateDetailsScreen(Rate rate) {
        startActivity(RateActivity.prepareStartIntent(this, rate.getCurrencyCode()));
    }

    @Override
    public ObservableField<RatesAdapter> getAdapter() {
        return ratesAdapter;
    }
}

File: kotlin/app/src/main/java/com/pgssoft/kotlinplayground/view/activity/MainActivity.kt
package com.pgssoft.kotlinplayground.view.activity

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.view.MenuItemCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.view.Menu
import com.pgssoft.kotlinplayground.KotlinApp
import com.pgssoft.kotlinplayground.R
import com.pgssoft.kotlinplayground.databinding.ActivityMainBinding
import com.pgssoft.kotlinplayground.model.db.Rate
import com.pgssoft.kotlinplayground.view.adapter.RatesAdapter
import com.pgssoft.kotlinplayground.view.adapter.providers.BaseItemProvider
import com.pgssoft.kotlinplayground.view.adapter.providers.RatesAdapterProvider
import com.pgssoft.kotlinplayground.viewmodel.MainActivityViewModel
import com.pgssoft.kotlinplayground.viewmodel.iface.IMainActivityAccess

class MainActivity : BaseViewModelActivity<MainActivityViewModel>(), IMainActivityAccess, RatesAdapterProvider {
    override lateinit var viewModel: MainActivityViewModel
    override lateinit var adapter: RatesAdapter

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val serviceProvider = KotlinApp.serviceProvider
        viewModel = MainActivityViewModel(this, serviceProvider.apiService, serviceProvider.repository)
        adapter = RatesAdapter(this, viewModel, viewModel)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        binding.recyclerView.layoutManager = LinearLayoutManager(this)

        binding.viewModel = viewModel
        binding.adapterProvider = this

        viewModel.onInfrastructureReady()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        val searchItem = menu.findItem(R.id.action_search)
        val searchView = MenuItemCompat.getActionView(searchItem) as SearchView

        searchView.setOnQueryTextListener(object: SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.filter(newText)
                return false
            }

            override fun onQueryTextSubmit(query: String?): Boolean {
                return false
            }
        })

        searchView.setOnCloseListener({
            viewModel.clearFiltering()
            false
        })

        return super.onCreateOptionsMenu(menu)
    }

    override fun displayRates(ratesProvider: BaseItemProvider<Rate>) {
        adapter.filter(ratesProvider)
    }

    override fun openRateDetailsScreen(rate: Rate) {
        startActivity(RateActivity.prepareStartIntent(this, rate.currencyCode))
    }
}
